code/controllers/subsystems/cleanup_ch.dm

```diff
@@ -0,0 +1,165 @@
+	var/list/old_processing = SScleanup.processing.Copy()
+	for(var/datum/D in old_processing)
+		if(istype(D))
+			processing |= D
+
+/datum/controller/subsystem/cleanup/stat_entry()
+	..("[stat_tag]:[processing.len] U:[urgency] P:[points]")
+
+/datum/controller/subsystem/cleanup/proc/get_points()
+	var/temp_points = processing.len
+	for(var/mob/mob_entry in processing)
+		if(istype(mob_entry))
+			temp_points += 2 //Mobs are worth 3x as much as atoms, since they're counted once in the first definition for temp_points, and twice here.
+	for(var/mob/living/carbon/carbon_entry in processing)
+		if(istype(carbon_entry))
+			temp_points += 2 //Carbons are worth 5x as much as atoms, since they're counted once in the first definition for temp_points, twice for being a mob, and twice for carbon.
+	return temp_points
+
+/proc/cleanup_adjusted_points(points)
+	return (log(points)-2.5)/3
+
+/datum/controller/subsystem/cleanup/proc/get_urgency() //Let's try to decide how badly we need to delete stuff. Urgency is always betweeen 0 and 1, 0 meaning no rush, and 1 meaning delete stuff twice as quick.
+	SScleanup.points = SScleanup.get_points()
+	var/adjusted_points = cleanup_adjusted_points(SScleanup.points) //Use the following points to get an idea for this equation: (12.182,0), (100,0.702), (244.692,1)
+	var/adjusted_tidi = cleanup_adjusted_points(2*SStime_track.time_dilation_avg) //We use average time dilation to avoid unwanted behavior during small lag spikes. Same equation as before, but x is multiplied by 2.
+	var/combined_score = adjusted_points * adjusted_tidi //We multiply them together, because we should only act urgently when both of these are above standard values.
+	var/out = SMOOTHSTEP(combined_score) //Finally we add smoothstep as a way of smoothly clamping between 0 and 1.
+	if(out > URGENCY_PANIC_MODE)
+		log_and_message_admins("WARNING: Cleanup subsystem has reached panic mode threshold, atoms and mobs in the queue will be deleted faster and regardless of location or last movement")
+	return out
+	// To further document what the math here is actually doing, the following are points in the form of (SScleanup.points, SStime_track.time_dilation_avg, urgency):
+	// (50,25,0.076), (50,50,0.205), (50,100,0.387), (100,25,0.205), (100,50,0.486), (100,100,0.772), (100,150,0.896), (150,25,0.307), (150,50,0.66), (150,100,0.926), (150,150, 0.99) 
+	
+
+
+/datum/controller/subsystem/cleanup/fire(resumed = 0)
+	if (!resumed)
+		currentrun = processing.Copy()
+		SScleanup.urgency = get_urgency()
+	var/list/current_run = currentrun
+	var/current_check = 0
+	while(current_run.len)
+		current_thing = current_run[current_run.len]
+		current_run.len--
+		if(QDELETED(current_thing) || QDELETED(current_thing.entry_atom))
+			processing -= current_thing
+		current_check = current_thing.Check()
```

Comment:
This is probably the biggest problem I see here. QDELETED returns true if the arg is null as well (because maybe it's already been GC'd).

In the case where the datum is null, this will just runtime and break the subsystem since you call Check() on it in the next line.

In the case where the atom is null, you will generate a runtime inside Check() below, which assumes the atom exists and calls CleanupCheck on it. This is much more likely, and I would imagine you'd encounter this probably somewhat often, maybe even once a shift: imagine a dead mob getting butchered, which would seemingly always trigger this.

You should probably `continue` inside this, after removing it from processing.

Reply:
This was me being a 4head, and a pretty big oversight lol. If I add continue here, I will probably want to also move the tick check to the beginning of the loop, so hmm, I will have to think about this. The alternative would just be adding an else statement and putting the rest of the code under that, and in the end it doesn't really matter other than for cleanliness.

Reply:
> This was me being a 4head, and a pretty big oversight lol. If I add continue here, I will probably want to also move the tick check to the beginning of the loop, so hmm, I will have to think about this. The alternative would just be adding an else statement and putting the rest of the code under that, and in the end it doesn't really matter other than for cleanliness.

I'm not sure I'd worry too much about that honestly. If you can think of an elegant way to do it, sure. But the 'worst case' for this is 'the entire processing list is QDELETED(x)=TRUE items' (so every single item processed goes into the inside of this if() statement), and I have trouble imagining that clearing the list out will be that expensive. 150 shell casings or whatever, aka removing 150 nulls from this list (just imagining that the shell casings were all destroyed by an explosion or something so they're all deleted and now null), should be baaaasically instant.

I guess if it gets into tracking like 100k+ things then I'd worry about the performance in the off-chance that you end up with 'large' runs of nulls in the list.

Reply:
> > This was me being a 4head, and a pretty big oversight lol. If I add continue here, I will probably want to also move the tick check to the beginning of the loop, so hmm, I will have to think about this. The alternative would just be adding an else statement and putting the rest of the code under that, and in the end it doesn't really matter other than for cleanliness.
> 
> I'm not sure I'd worry too much about that honestly. If you can think of an elegant way to do it, sure. But the 'worst case' for this is 'the entire processing list is QDELETED(x)=TRUE items' (so every single item processed goes into the inside of this if() statement), and I have trouble imagining that clearing the list out will be that expensive. 150 shell casings or whatever, aka removing 150 nulls from this list (just imagining that the shell casings were all destroyed by an explosion or something so they're all deleted and now null), should be baaaasically instant.
> 
> I guess if it gets into tracking like 100k+ things then I'd worry about the performance in the off-chance that you end up with 'large' runs of nulls in the list.
> 
> EDIT: I definitely wouldn't move the tickcheck to the top though, since that could result in the subsystem not doing ANY work during superlag times where the MC tells the subsystem it's out of time already as it's starting. It should do at least 1 item per run, so if everything is going to hell, it can at least slowly work on it.
> 
> EDIT EDIT: Actually yeah, wait, if everything is going to hell and lagging, don't you want to NOT stop working if the MC says it's real laggy? The whole TICK CHECK stuff is to give up working to allow other things time to do their jobs when it's laggy times, but like. If the server is turbo-lagged don't you want to ESPECIALLY do work with this subsystem then? HMM.

Hmm. I'll think about this, I could keep it at the end and then use an urgency check to decide whether or not to keep going, but even then it will probably only be up to like 5 entries per tick, because it should still keep firing every tick afterwards until it gets through the queue.